### PR TITLE
Added 'add_data_validation` to `Workhsheet` [Issue #1420]

### LINKED ISSUE #1420: A way to add DataValidation
**Is your feature request related to a problem? Please describe.**
The API allows to setup data validation on a range via the `batchUpdate` end point: https://developers.google.com/sheets/api/samples/data#data-validation

**Describe the solution you'd like**
Since it is part of the `batchUpdate` endpoint, anyone can use the low-level method to pass through the data. However, there should be a high-level method. or a kwarg in the low-level [`values_update`](https://github.com/burnash/gspread/blob/4a48fbdc4bc56724d3a1d8fcdb7f5793fb58f160/gspread/http_client.py#L138) method.  We should create a class `DataValidation` class, which will have all the options provided by the API.

It would have `to_json` method, that will be passed to the update payload when sending the request.

**Describe alternatives you've considered**
Passing a `kwarg` to the update methods to pass through json data.

### PR DESCRIPTION
**Pull request by muddi900**
Added a RealtiveDate enum

removed the realtive date enum

Closes #1420

### CODE REVIEW

**Comment by alifeee** [comment]
hi :) thanks for the PR

it looks like you have to run `tox -e format` locally to format the code, so the linter is happy.

after that, the CI should pass, and we will be able to review this change :)

**Comment by alifeee** [comment]
Hi, that looks good. Sorry there keep being issues.

Try deleting the cassette, and re-running the test to generate it again. Sometimes there are conflicts.

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3211]: Please remove the `kwargs` and use explicit arguments: `inputMessage`, `strict`, `showCustomUi`.

We tend to remove most kwargs and prevent using them when not necessary.

File: gspread/utils.py
Comment (lavigne958) [inline gspread/utils.py:121]: Could you add a new enum too for `RelativeDate` possible values ?

Documented here: https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/other#relativedate

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3225]: Can we please add a couple or examples in the documentation like: `NUMBER_GREAtER`.

And/or (depending on the time to make it, a popular based condition like : ONE_OF_LIST

**Review by lavigne958 — COMMENTED** [review]
Hi thank you for this PR, I need to test it and check the API to confirm.

**Comment by lavigne958** [comment]
What type of cassette recording mode did you use ?

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3240]: Here we should check that the condition type is part of the enum otherwise we should raise an error.

File: gspread/utils.py
Comment (muddi900) [reply gspread/utils.py:121]: I originally did add an enum for Relative Date but it is not supported by validation requests.

File: gspread/worksheet.py
Comment (muddi900) [reply gspread/worksheet.py]: Added in 811ff02b40a53757ca52d259b5b6ced59106365f

File: gspread/worksheet.py
Comment (muddi900) [reply gspread/worksheet.py]: Added in 06c3f24113244b1b8a5f285e5b5976db396d3a22

File: gspread/worksheet.py
Comment (muddi900) [reply gspread/worksheet.py:3275]: added in 811ff02b40a53757ca52d259b5b6ced59106365f

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3213]: This syntax won't work on earliest versions of python we support. please use `Optional[str]` instead

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3212]: I don't fancy having an unbounded list of arguments, it has been quiet complex to simplify this project and add typing.

I would prefer to have a single argument that hold either a list, a tuple etc (any iterable in fact) and keep the number of arguments fixed in size. 

@alifeee  what do you think about this ?

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3236]:
```suggestion
        **Examples**

        .. code-block:: python

            import gspread
```
you must add a blank line before and after the code-block

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3223]:
```suggestion
            .. _ConditionType: https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/other#ConditionType
```
you need a space between the start `..` and the URL reference `_ConditionType`

**Review by lavigne958 — COMMENTED** [review]
thank you for this proposal. it's well made.

You do need to change the typing notation in order to pass the CI linter.

File: gspread/worksheet.py
Comment (muddi900) [reply gspread/worksheet.py]: I was going back and forth between unbouded values and an iterable. I have changed it in a35c39fbf80ae6b866f0eb38cda73c131666c4e1

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3210]: could you please rename the argument `source` to `range`, we intend to use `range` for range name.

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3213]: I don't understand why is this unnamed argument here then if we have an actual list of values into a single argument.

File: gspread/worksheet.py
Comment (muddi900) [reply gspread/worksheet.py:3210]: I get lint errors when I do that.

Comment (lavigne958) [reply gspread/worksheet.py:3210]: I think we missunderstood each others, what I mean is: the method argument currently name `source` should be renamed `range` please.
This is a method argument name I don't see how it can fail the linter :thinking:  unless there is something I don't see here.

File: gspread/worksheet.py
Comment (lavigne958) [inline gspread/worksheet.py:3238]: you do need a blank line here between `.. code-block:: python` and the first line of code bellow in order to pass documentation build step

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py:3213]: I checked-out your branch, and the tests are failing because of the `*` argument.

you need to remove it.
currently the arguments:
- self
- ~source~ range
- condition_type
- values

are regular args with no default value.

The arguments:
- inputMessage
- strict
- showCustomUi

are kwargs, because they have a default value set.

**Review by lavigne958 — COMMENTED** [review]
Please follow the given indications so the test + documentation build can succeed. 

Could you please rename the argument `showCustomUi` to `showCustomUI` with a uppercase `i`, it's an acronym so it should be uppercase. 

after that it's all good to, we will merge it once the changes are applied.

**Comment by lavigne958** [comment]
I checked and the tests won't pass until you remove the following file and record it again.
`tests/cassettes/WorksheetTest.test_add_validation.json`

Quick trick, in order to regenerate only this cassette you can use the following command:
`tox -e py -- -k test_add_validation`
don't forget to set the env variables `GS_CREDS_FILENAME` and `GS_RECORD_MODE=new_episodes` so it only records this missing cassette.

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py:3237]: This has not been resolved. Please add a blank line before you start your code lines.

**Comment by lavigne958** [comment]
Great thanks. Last thing left is the documentation to fix then all tests should pass.

**Comment by muddi900** [comment]
Seems all tests are working, I'll flatten the commits in a moment

**Review by lavigne958 — APPROVED** [review]

File: gspread/worksheet.py
Comment (alifeee) [inline gspread/worksheet.py:3258]: does this example work with the new format of [values]?

Comment (lavigne958) [reply gspread/worksheet.py:3258]: good catch  ! no it does not in fact, values must be put into a single list.

**Comment by alifeee** [comment]
this looks wonderful

would we like to put it in any docs like the main `rst` example file, or the README, so that people actually know this new function exists?

:)

**Comment by muddi900** [comment]
If there is a markdown to rst converter, I'll be game.

**Comment by alifeee** [comment]
hmm, not sure about markdown to `rst`...

but I would add a simple code snippet, like
```python
ws.add_validation(
                'C2:C7',
                ValidationConditionType.one_of_list,
                ['Yes','No'],
                showCustomUi=True
            )
```

to the README (1)...

<https://github.com/burnash/gspread/blob/master/README.md?plain=1>

...and the user guide (2)...

<https://github.com/burnash/gspread/blob/master/docs/user-guide.rst?plain=1>

... in an appropriate place. For (1), it's markdown. For (2), you can run `tox -e doc` to build the docs to see if they build properly.

File: tests/worksheet_test.py
Comment (alifeee) [inline tests/worksheet_test.py:1883]: like the docs, this is also wrong now, no? it is only luck that it passes

it should be ["y", "n"]

it may also be nice to add one other validation condition to the tests, like `text_contains`, or `number_not_between`, etc, to get a bit more coverage.

(I think there is also an argument for testing each type of condition, since we *are* the ones to be providing each `enum`)... @lavigne958 ?

File: README.md
Comment (lavigne958) [inline README.md:97]: Why is this changing please ?

File: README.md
Comment (lavigne958) [inline README.md:281]: here too why is this changing please ?

**Review by lavigne958 — COMMENTED** [review]
great addon with the documentation.

please remove unwanted changes and we should be good to go

**Review by lavigne958 — APPROVED** [review]

**Comment by lavigne958** [comment]
great work thank you for this contribution !

**Pull request merged**
